File: 1056-capacity-to-ship-packages-within-d-days/1056-capacity-to-ship-packages-within-d-days.py
class Solution:
    def isPossible(self, arr, n, days, mid):
        load = 0
        used = 1
        for i in range(0, n):
            if load + arr[i] <= mid:
                load += arr[i]
            else:
                load = arr[i]
                used += 1
        
        if used > days:
            return False
        else:
            return True

    def shipWithinDays(self, arr: List[int], days: int) -> int:
        n = len(arr)
        st = max(arr)
        end = sum(arr)
        ans = 0  # Initialize ans here
        
        while st <= end:
            mid = st + (end - st) // 2
            if self.isPossible(arr, n, days, mid):
                ans = mid
                end = mid - 1
            else:
                st = mid + 1
        return ans

File: 1056-capacity-to-ship-packages-within-d-days/1056-capacity-to-ship-packages-within-d-days.cpp
class Solution {
public:

    bool isPossible(vector<int> &arr, int n, int days, int mid)
    {
        int load = 0, used = 1;
        for(int i = 0; i<n; i++)
        {
            if(load+arr[i] <= mid)
            {
                load += arr[i];
            }else
            {
                used++;
                load = arr[i];
            }
        }
        return used > days ? false : true;
    }

    int shipWithinDays(vector<int>& arr, int days) {
        int n = arr.size();
        int st = *max_element(arr.begin(), arr.end());
        int sum = 0;
        for(int i : arr) sum += i;
        int end = sum;
        int ans = 0;
        while(st <= end)
        {
            int mid  = st + (end - st) / 2;
            if(isPossible(arr, n, days, mid))
            {
                ans = mid;
                end = mid - 1;
            }else
            {
                st = mid + 1;
            }
        }
        return ans;
    }
};